session_sync_test.go

```diff
@@ -0,0 +1,163 @@
+package remotedialer
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	"net/http/httptest"
+	"reflect"
+	"sort"
+	"testing"
+
+	"github.com/gorilla/websocket"
+)
+
+func Test_encodeConnectionIDs(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		size int
+	}{
+		{0},
+		{1},
+		{2},
+		{10},
+		{100},
+		{1000},
+	}
+	for _, tt := range tests {
```

Comment:
You'd need to capture the loop variable to avoid the problem with the same variable would be used in closures when multiple goroutines are present.

```suggestion
	for _, tt := range tests {
	      tt := tt
```

https://eli.thegreenplace.net/2019/go-internals-capturing-loop-variables-in-closures/

Reply:
Wow, I completely missed it, thanks. This caveat is changing in 1.22 though:
https://go.dev/blog/loopvar-preview
I have that version locally:
```
❯ go test -v -run 'Test_diffSortedSetsGetRemoved.*' ./
=== RUN   Test_diffSortedSetsGetRemoved
=== PAUSE Test_diffSortedSetsGetRemoved
=== CONT  Test_diffSortedSetsGetRemoved
=== RUN   Test_diffSortedSetsGetRemoved/case_0
=== PAUSE Test_diffSortedSetsGetRemoved/case_0
=== RUN   Test_diffSortedSetsGetRemoved/case_1
=== PAUSE Test_diffSortedSetsGetRemoved/case_1
=== RUN   Test_diffSortedSetsGetRemoved/case_2
=== PAUSE Test_diffSortedSetsGetRemoved/case_2
=== RUN   Test_diffSortedSetsGetRemoved/case_3
=== PAUSE Test_diffSortedSetsGetRemoved/case_3
=== CONT  Test_diffSortedSetsGetRemoved/case_0
=== CONT  Test_diffSortedSetsGetRemoved/case_2
=== CONT  Test_diffSortedSetsGetRemoved/case_1
=== CONT  Test_diffSortedSetsGetRemoved/case_3
--- PASS: Test_diffSortedSetsGetRemoved (0.00s)
    --- PASS: Test_diffSortedSetsGetRemoved/case_0 (0.00s)
    --- PASS: Test_diffSortedSetsGetRemoved/case_2 (0.00s)
    --- PASS: Test_diffSortedSetsGetRemoved/case_1 (0.00s)
    --- PASS: Test_diffSortedSetsGetRemoved/case_3 (0.00s)
PASS
ok  	github.com/rancher/remotedialer	0.209s
```
but it still not used in this project.

Having said that, I'm not sure it's worth the "complexity", just for running these tests in parallel. Locally for me, they used to always run under 1 second, and using `Parallel()` and `-p=10` or `20` does not have any noticeable impact.
I'm keen to revert that change and keep them simpler, since I believe these unit tests won't benefit much from running in parallel. What do you think?

Reply:
Even in Go 1.22, this old behavior is observed. Basically, the last subtest will be run twice. So if the last subtest passes, it would appear as if the whole test passes, even if previous subtests fail. This behavior hides test failures.
In Go 1.22, you can forego the capturing of the loop variable and have the expected behavior if you use the experimental environment variable `GOEXPERIMENT=loopvar go test`. Maybe I am missing something, but it's not the default yet.

It's always good to run tests and subtests in parallel because Go would run parallel tests together with **other** parallel tests - that's the main benefit. It's not about running these individual tests in parallel in isolation. If your functions under test took more time (you can simulate this with `time.Sleep` calls), you'd see the difference in running them sequentially vs. in parallel vs. in parallel with parallel subtests.

Of course, tests and functions under test need to be eligible to be run in parallel, and that's possible when there is no global state shared among them. So running in parallel is a safe default unless there is a specific reason not to do this.

Reply:
Oh, you are right, the subtest name is created outside the closure, that's why it's correct.

OK, I'll keep the parallel marks, no strong opinion from my side.

Reply:
> Even in Go 1.22, this old behavior is observed.

From the link I posted:
> To ensure backwards compatibility with existing code, the new semantics will only apply in packages contained in modules that declare go 1.22 or later in their go.mod files.

Reply:
Ah, you're right! So if the module declares that it uses 1.20, it doesn't matter that my local toolchain is 1.22, makes sense, thanks.